compiler/src/lnntoamm/Types.ts

```diff
@@ -123,127 +193,279 @@ export default abstract class Type implements Equalable {
+    if (that instanceof Opaque) {
+      if (
+        Object.values(this.generics).some((g) => g === null) ||
+        Object.values(that.generics).some((g) => g === null)
+      ) {
+        // if any values are null values, that means that we just have to check
+        // for constraint compatibilities which was already checked so we're good
+        return;
+      }
+      const thisGens = Object.keys(this.generics);
+      const thatGens = Object.keys(that.generics);
+      thisGens.forEach((genName, ii) =>
+        this.generics[genName].constrain(that.generics[thatGens[ii]], scope),
       );
+    } else if (that instanceof Interface || that instanceof OneOf) {
+      that.constrain(this, scope);
+    } else {
+      console.log(this);
+      console.log(that);
+      throw 'uh';
     }
   }
 
-  eq(that: Equalable): boolean {
-    if (that instanceof Builtin) {
-      return this === that;
-    } else if (that instanceof Interface) {
-      if (that instanceof Generated && that.delegate !== null) {
-        return this.eq(that.delegate);
+  contains(that: Type): boolean {
+    return this.eq(that);
+  }
+
+  dup(): Type | null {
+    const genKeys = Object.keys(this.generics);
+    if (genKeys.length === 0) {
+      return null;
+    }
+    const duped = new Opaque(this.name, genKeys);
+    let isNothingNew = true;
+    genKeys.forEach((genName) => {
+      const thisGen = this.generics[genName];
+      let tyVal: Type;
+      if (thisGen === null) {
+        tyVal = Type.generate();
+      } else {
+        const duped = thisGen.dup();
+        if (duped === null) {
+          tyVal = thisGen;
+        } else {
+          tyVal = duped;
+          isNothingNew = false;
+        }
       }
-      return that.tempDelegate !== null && this.eq(that.tempDelegate);
-    } else {
+      duped.generics[genName] = tyVal;
+    });
+    return duped;
+  }
+
+  eq(that: Equalable): boolean {
+    if (!(that instanceof Opaque) || this.name !== that.name) {
       return false;
     }
+    const thisGens = Object.values(this.generics);
+    const thatGens = Object.values(that.generics);
+    return (
+      thisGens.length === thatGens.length &&
+      thisGens.every((thisGen, ii) => {
+        const thatGen = thatGens[ii];
+        if (thisGen === null || thatGen === null) {
+          return thisGen === thatGen;
+        } else {
+          return thisGen.eq(thatGen);
+        }
+      })
+    );
   }
 
-  fieldIndices(): FieldIndices {
-    return TODO("determine if it's even worth keeping the Builtin class");
+  fnselectOptions(): Type[] {
+    const genOptions = Object.values(this.generics).map((g) =>
+      g === null ? [g] : g.fnselectOptions(),
+    );
+    const opts = new Array<Type>();
+    const getIndices = matrixIndices(genOptions);
+    const toSolidify = new Opaque(this.name, Object.keys(this.generics));
+    for (
+      let indicesRes = getIndices.next();
+      !indicesRes.done;
+      indicesRes = getIndices.next()
+    ) {
```

Comment:
This looks like an iterator, couldn't it be refactored to:

```suggestion
    for (let indicesRes of getIndices) {
```

Reply:
it's a function generator - there might be a more convenient way to do this but I don't believe `let of` works with generators

Reply:
ah, `for of` does work with generators <https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for...of#iterating_over_generators>